File: Top_Interview_150/Array_String/cpp/380.cpp
#include <vector>
#include <unordered_map>
#include <random>

using namespace std;

class RandomizedSet
{
    unordered_map<int, int> hash_idx;
    vector<int> data;
    random_device rd;
    mt19937 gen;

public:
    RandomizedSet()
    {
    }

    bool insert(int val)
    {
        if (this->hash_idx.find(val) == hash_idx.end())
        {
            this->data.push_back(val);
            this->hash_idx[val] = data.size() - 1;
            return true;
        }
        return false;
    }

    bool remove(int val)
    {
        auto idx_data = this->hash_idx.find(val);
        if (idx_data != this->hash_idx.end())
        {
            int idx = idx_data->second;
            this->data[idx] = this->data.back();
            this->hash_idx[this->data.back()] = idx;

            this->data.pop_back();
            this->hash_idx.erase(val);
            return true;
        }
        return false;
    }

    int getRandom()
    {
        uniform_int_distribution<> distrib(0, data.size() - 1);
        return data[distrib(gen)];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: Top_Interview_150/Array_String/python/380.py
from random import choice


class RandomizedSet:

    def __init__(self):
        self.hash_lst = dict()
        self.lst = list()

    def insert(self, val: int) -> bool:
        if val not in self.hash_lst:
            self.lst.append(val)
            self.hash_lst[val] = len(self.st) - 1
            return True

        return False

    def remove(self, val: int) -> bool:
        if val in self.hash_lst:
            idx = self.hash_lst[val]
            self.lst[idx] = self.lst[-1]
            self.hash_lst[self.lst[-1]] = idx
            self.lst.pop()

            del self.hash_lst[val]
            return True

        return False

    def getRandom(self) -> int:
        return choice(self.lst)


# Your RandomizedSet object will be instantiated and called as such:
